### PR TITLE
Replace with unix timestamp

### PR DESCRIPTION
**Pull request by FredrikAreschoug**

### CODE REVIEW

File: src/main/java/com/hedvig/notificationService/customerio/UnixTimestampExtension.kt [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+        when (val value = it.value) {
+            is LocalDate -> mutableMap[it.key] = value.localDateToUnixTimestamp()
+            is LocalDateTime -> mutableMap[it.key] = value.toEpochSecond(ZoneOffset.UTC)
+            is Instant -> mutableMap[it.key] = value.atZone(ZoneOffset.UTC).toEpochSecond()
+            is String -> {
+                try {
+                    mutableMap[it.key] = LocalDate.parse(value).localDateToUnixTimestamp()
+                } catch (e: Exception) {
+                    // no-op
+                }
```
Comment (johantjellden) [inline src/main/java/com/hedvig/notificationService/customerio/UnixTimestampExtension.kt:21]: We should probably have a case where we need to go deeper if the value is a new map.

File: src/test/java/com/hedvig/notificationService/utils/UnixTimestampExtensionTest.kt [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+import org.junit.jupiter.api.extension.ExtendWith
+import java.time.Instant
+import java.time.LocalDate
+import java.time.LocalDateTime
+
+@ExtendWith(MockKExtension::class)
+class UnixTimestampExtensionTest {
+
+    @Test
+    fun `given LocalDate of sign date is replace with unix timestamp`() {
```
Comment (johantjellden) [inline src/test/java/com/hedvig/notificationService/utils/UnixTimestampExtensionTest.kt:17]: I think we need a test like this:
```
val map: Map<String, Any?> = mapOf("member_id" to "1234", "signed_at" to LocalDateTime.of(2020, 8, 17, 20, 17))

val dataMap: Map<String, Any?> = mapOf("data" to map)

val mapWithUnixTimestamp = map.replaceWithUnixTimestamp()

assertThat(mapWithUnixTimestam["data"]["signed_at"]). .isEqualTo(1597695420L)
```

File: src/main/java/com/hedvig/notificationService/customerio/UnixTimestampExtension.kt [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+                @Suppress("UNCHECKED_CAST")
+                mutableMap[it.key] = (value as Map<String, Any?>?)?.replaceWithUnixTimestamp()
+            }
+        }
+    }
+
+    return mutableMap.toMap()
+}
+
+private fun LocalDate.localDateToUnixTimestamp() = this.toEpochSecond(LocalTime.of(0, 0), ZoneOffset.UTC)
```
Comment (johantjellden) [inline src/main/java/com/hedvig/notificationService/customerio/UnixTimestampExtension.kt:33]: We should pass in the members timezone here in order to properly convert it back to the local date in customer.io.

Consider this example:
```
val now = LocalDate.now() // java.time.LocalDate! = 2020-11-09
val instant = now.toEpochSecond(LocalTime.of(0, 0), ZoneOffset.UTC) // kotlin.Long = 1604880000
Instant.ofEpochSecond(instant).atZone(ZoneId.of("America/New_York")).toLocalDate() // java.time.LocalDate! = 2020-11-08
Instant.ofEpochSecond(1604880000).atZone(ZoneId.of("Europe/Reykjavik")).toLocalDate()
```

Comment (FredrikAreschoug) [reply src/main/java/com/hedvig/notificationService/customerio/UnixTimestampExtension.kt:33]: Don't we want it in UTC in customer.io? or do we store the customers time zone in cutomer.io?

File: src/main/java/com/hedvig/notificationService/customerio/CustomerioService.kt [hunk 4]
```diff
@@ -47,7 +47,7 @@ class CustomerioService(
             return
         }
 
-        clients[marketForMember]?.updateCustomer(memberId, attributes)
+        clients[marketForMember]?.updateCustomer(memberId, attributes.replaceWithUnixTimestamp(marketForMember.zoneId))
```
Comment (johantjellden) [inline src/main/java/com/hedvig/notificationService/customerio/CustomerioService.kt:50]: One last thing, lets add a feature flag for this functionality.
For this I would suggest using a property and not a profile.

Comment (FredrikAreschoug) [reply src/main/java/com/hedvig/notificationService/customerio/CustomerioService.kt:50]: What will this give us?

**Pull request merged**